pkg/cache/v3/linear.go

```diff
@@ -29,59 +30,74 @@ import (
 // cachedResource is used to track resources added by the user in the cache.
 // It contains the resource itself and its associated version (currently in two different modes).
 type cachedResource struct {
-	types.Resource
+	name string
+
+	resource types.Resource
+	ttl      *time.Duration
 
 	// cacheVersion is the version of the cache at the time of last update, used in sotw.
 	cacheVersion string
-	// stableVersion is the version of the resource itself (a hash of its content after deterministic marshaling).
-	// It is lazy initialized and should be accessed through getStableVersion.
-	stableVersion string
 }
 
-func newCachedResource(res types.Resource, cacheVersion string) *cachedResource {
+func newCachedResource(name string, res types.Resource, cacheVersion string) *cachedResource {
 	return &cachedResource{
-		Resource:     res,
+		name:         name,
+		resource:     res,
 		cacheVersion: cacheVersion,
 	}
 }
 
-func (c *cachedResource) getStableVersion() (string, error) {
-	if c.stableVersion != "" {
-		return c.stableVersion, nil
+func newCachedResourceWithTTL(name string, res types.ResourceWithTTL, cacheVersion string) *cachedResource {
+	return &cachedResource{
+		name:         name,
+		resource:     res.Resource,
+		ttl:          res.TTL,
+		cacheVersion: cacheVersion,
 	}
+}
 
-	// TODO(valerian-roche): store serialized resource as part of the cachedResource
-	// to reuse it when marshaling the responses instead of remarshaling and recomputing the version then.
-	marshaledResource, err := MarshalResource(c.Resource)
-	if err != nil {
-		return "", err
-	}
-	c.stableVersion = HashResource(marshaledResource)
-	return c.stableVersion, nil
+// getMarshaledResource lazily marshals the resource and returns the bytes.
+func (c *cachedResource) getMarshaledResource() ([]byte, error) {
+	return sync.OnceValues(func() ([]byte, error) {
```

Comment:
I'm not sure to understand how this effectively synchronizes the calls here. If you call the function returned by `sync.OnceValues`, then different calls to `getMarshaledResource` are not synchronized. AFAICT, the call to `sync.OnceValues` is either unnecessary or not doing its job.

Reply:
Fixed. Missed the implementation part. I did a quick test to confirm it's now serialized only once